modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordWaitlistDialog.kt

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * `java.net.http.HttpRequest.BodyPublishers` does not have a dedicated publisher for the multipart forms.
+ */
+private fun HttpRequest.Builder.postMultipartFormData(data: Iterable<Pair<String, String>>): HttpRequest.Builder {
+    val boundary = UUID.randomUUID().toString()
+    val byteArrays = mutableListOf<ByteArray>()
+    val separator = "--$boundary\r\nContent-Disposition: form-data; name=".toByteArray(StandardCharsets.UTF_8)
+
+    data.forEach {
+        byteArrays.add(separator)
+        byteArrays.add("\"${it.first}\"\r\n\r\n${it.second}\r\n".toByteArray(StandardCharsets.UTF_8))
+    }
+
+    byteArrays.add("--$boundary--".toByteArray(StandardCharsets.UTF_8))
+
+    this
+        .header("Content-Type", "multipart/form-data;boundary=$boundary")
+        .POST(HttpRequest.BodyPublishers.ofByteArrays(byteArrays))
+
+    return this
+}
+
+class MirrordWaitlistDialog(private val project: Project) {
+    /**
+     * Background task for making the signup request.
+     */
+    private class SignupTask(project: Project, private val email: String) : Task.Backgroundable(project, "mirrord", true) {
+        private val service: MirrordProjectService = project.service()
+
+        override fun run(indicator: ProgressIndicator) {
+            indicator.text = "making the signup request..."
+
+            val client = HttpClient.newHttpClient()
+            val request = HttpRequest
+                .newBuilder(URI(SIGNUP_ENDPOINT))
+                .timeout(Duration.ofSeconds(5))
+                .postMultipartFormData(listOf(Pair("email", email)))
+                .build()
+
+            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
+            if (response.statusCode() != 200) {
+                throw RuntimeException("invalid response status ${response.statusCode()}")
+            }
+        }
+
+        override fun onThrowable(error: Throwable) {
+            MirrordLogger.logger.debug("waitlist signup failed", error)
+            service.notifier.notifyRichError("Failed to join the waitlist: ${error.message}")
+        }
+
+        override fun onSuccess() {
+            service.notifier.notifySimple(
+                "Thank you for joining the waitlist for mirrord for Teams! We'll be in touch soon.",
+                NotificationType.INFORMATION
+            )
+        }
+    }
+
+    /**
+     * Display a simple dialog with email input.
+     * Waitlist signup is requested when the user presses the "Ok" button.
+     */
+    fun show() {
+        val dialog = DialogBuilder()
+
+        val textField = JTextField().apply {
+            val field = this
+            document.addDocumentListener(object : DocumentListener {
+                override fun insertUpdate(e: DocumentEvent) = doUpdate()
+                override fun removeUpdate(e: DocumentEvent) = doUpdate()
+                override fun changedUpdate(e: DocumentEvent) = doUpdate()
```

Comment:
Are all three necessary? I would think you'd either the first two or the last one

Reply:
I'm not really sure, that's what I saw in the target selection dialog. I don't think it matters though, there is no heavy computation in `doUpdate`